modules/installation-osp-local-disk-deployment.adoc

```diff
@@ -0,0 +1,296 @@
+      "value": [
+        {
+          "name": "etcd",
+          "sizeGiB": 10,
+          "storage": {
+            "type": "Local" <3>
+          }
+        }
+      ]
+    }
+  ]
+'
+----
+<1> Applies the JSON patch to the `ControlPlaneMachineSet` CR.
+<2> Specifies the path where the `additionalBlockDevices` are added.
+<3> Adds the `etcd` devices with local storage of 10 GB to the cluster.
+====
+
+. Verify that the control plane machines are healthy:
+
+.. Verify the cluster is progressing by running the following command:
+
++
+[source,terminal]
+----
+$ oc wait --timeout=90m --for=condition=Progressing=false controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
+----
+
+.. Verify that 3 replica control plane machine sets are running by running the following command:
+
++
+[source,terminal]
+----
+$ oc wait --timeout=90m --for=jsonpath='{.spec.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
+----
+
+.. Verify that the 3 control plane machine sets are updated by running the following command:
+
++
+[source,terminal]
+----
+$ oc wait --timeout=90m --for=jsonpath='{.status.updatedReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
+----
+
+.. Verify that the 3 control plane machine sets are healthy by running the following command:
+
++
+[source,terminal]
+----
+$ oc wait --timeout=90m --for=jsonpath='{.status.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
+----
+
+.. Verify the `ClusterOperators` are progressing in the cluster by running the following command:
+
++
+[source,terminal]
+----
+$ oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false
+----
+
+.. Verify that each of the 3 control plane machines has the additional block device you previously created by running the following script:
+
++
+[source,bash]
+----
+$ cp_machines=$(oc get machines -n openshift-machine-api --selector='machine.openshift.io/cluster-api-machine-role=master' --no-headers -o custom-columns=NAME:.metadata.name) <1>
+
+
+if [[ $(echo "${cp_machines}" | wc -l) -ne 3 ]]; then
+  exit 1
+fi <2>
+
+
+for machine in ${cp_machines}; do
+  if ! oc get machine -n openshift-machine-api "${machine}" -o jsonpath='{.spec.providerSpec.value.additionalBlockDevices}' | grep -q 'etcd'; then
+	exit 1
+  fi <3>
+done
+----
+<1> Retrieves the machines running in the cluster.
+<2> Checks that there are 3 machines running.
+<3> Checks that 3 `etcd` devices are running on the 3 machines.
+
+. Create a file named `98-var-lib-etcd.yaml` by using the following content:
+
+[WARNING]
+====
+This procedure can testing etcd on local disk and should not be used on production cluster because in certain cases complete loss of the control plane can occur. For more information, see "Overview of backup and restore operation" under "Backup and restore".
+====
+
```

Comment:
I wonder if we should document what the configuration below does. How about:

The following configuration will reformat the ephemeral disk created by Nova, mount it to /var/lib/etcd, and copy any previous etcd database to the new mountpoint before starting etcd. Nova creates a filesystem with the label 'ephemeral0'. We look for the existence of this filesystem and reformat it with the label 'local-etcd' if we find it.

It uses the following systemd units:

#### var-lib-etcd.mount

This defines the mountpoint and the device which will be mounted there. Because we are relying on filesystem labels, we specify the device as /dev/disk/by-label/local-etcd. This link is created automatically by udev. The mount unit has an implicit dependency on the device unit. We use this dependency to trigger the create-local-etcd.service unit below. We also rely on create-local-etcd.service to fail quickly if the device can't be created. This avoids the mount unit hanging waiting for the device to exist.

#### create-local-etcd.service

This runs as a dependency of the /dev/disk/by-label/local-etcd device. It formats the ephemeral0 device with an xfs filesystem and the label 'local-etcd'. It is skipped if the device already exists, which ensures we don't format a device with data on it.

Note that this unit intentionally fails if the current machine does not have an etcd ephemeral disk. This ensures that this configuration can be safely applied before adding ephemeral disks, or that the control plane machineset can be safely updated to remove the etcd ephemeral disk without first removing this configuration. This ensures that the /var/lib/etcd mount unit will not wait indefinitely for it to be created, and systemd will proceed immediately without the mount point. A side effect will be that systemd will list this unit as failed in that case. This is expected behaviour. To resolve this, remove this configuration after ensuring no control plane machine is using an etcd ephemeral disk.

#### migrate-to-local-etcd.service

This copies any existing local etcd data to the new mountpoint. It only runs if there is existing etcd data on the root device, but not in the new mount point.

#### relabel-var-lib-etcd.service

This ensures that the etcd mount point and any data migrated from the root device have the correct SELinux labels before etcd starts.